components-rs/bytes.rs

```diff
@@ -0,0 +1,227 @@
+        return object.$field.contains_key(&key);
+    }};
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_service_(ptr: *mut SpanBytes, str: *mut ZendString) {
+    set_string_field!(ptr, str, service);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_name_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, name);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_resource_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, resource);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_type_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, r#type);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(ptr, key, convert_to_bytes_string(val), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, key, meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_get_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+) -> *mut c_char {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let span = &mut *ptr;
+
+    let key = convert_to_bytes_string(key);
+
+    match span.meta.get(&key) {
+        Some(value) => {
+            let cstring = match CString::new(value.as_str()) {
+                Ok(s) => s,
+                Err(_) => CString::new("").unwrap_or_default(),
+            };
+            cstring.into_raw()
+        }
+        None => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_has_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) -> bool {
+    exists_hashmap!(ptr, key, meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_metrics_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: f64,
+) {
+    insert_hashmap!(ptr, key, val, metrics);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_metrics_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, key, metrics);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_get_span_metrics_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    result: *mut f64,
+) -> bool {
+    if ptr.is_null() {
+        return false;
+    }
+
+    let span = &mut *ptr;
+
+    let key = convert_to_bytes_string(key);
+
+    match span.metrics.get(&key) {
+        Some(&value) => {
+            *result = value;
+            true
+        }
+        None => false,
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_has_span_metrics_(ptr: *mut SpanBytes, key: *mut ZendString) -> bool {
+    exists_hashmap!(ptr, key, metrics);
+}
+
+pub unsafe extern "C" fn ddog_add_span_meta_struct_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(ptr, key, convert_to_bytes(val), meta_struct);
+}
```

Comment:
Let's also add a variant taking a CharSlice as key. Many of the keys are hardcoded and don't exist as zend_string.
It's less expensive too, to use `impl From<&'static str> for BytesString { fn from(value: &'static str) -> Self {` that function.
`key.as_slice().into()`, then.